packages/protocol/contracts/stability/Broker.sol

```diff
@@ -0,0 +1,205 @@
+    return true;
+  }
+
+  /**
+   * @notice Set the Mento reserve address
+   * @param _reserve The Mento reserve address
+   */
+  function setReserve(address _reserve) public onlyOwner {
+    require(_reserve != address(0), "Reserve address must be set");
+    emit ReserveSet(_reserve, address(reserve));
+    reserve = IReserve(_reserve);
+  }
+
+  /**
+   * @notice Calculate amountIn of tokenIn for a given amountIn of tokenIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountOut The amount of tokenOut to be bought
+   * @return amountIn The amount of tokenIn to be sold
+   */
+  function getAmountIn(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountOut
+  ) external returns (uint256 amountIn) {
+    require(isExchangeManager[exchangeManager], "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).getAmountIn(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountOut
+      );
+  }
+
+  /**
+   * @notice Calculate amountOut of tokenOut for a given amountIn of tokenIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountIn The amount of tokenIn to be sold
+   * @return amountOut The amount of tokenOut to be bought
+   */
+  function getAmountOut(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn
+  ) external returns (uint256 amountOut) {
+    require(isExchangeManager[exchangeManager], "ExchangeManager does not exist");
+    return
+      amountOut = IExchangeManager(exchangeManager).getAmountOut(
+        exchangeId,
+        tokenIn,
+        tokenOut,
+        amountIn
+      );
+  }
+
+  /**
+   * @notice Execute a token swap with fixed amountIn
+   * @param exchangeManager the address of the exchange manager for the pair
+   * @param exchangeId The id of the exchange to use
+   * @param tokenIn The token to be sold
+   * @param tokenOut The token to be bought
+   * @param amountIn The amount of tokenIn to be sold
+   * @param amountOutMin Minimum amountOut to be received - controls slippage
+   * @return amountOut The amount of tokenOut to be bought
+   */
+  function swapIn(
+    address exchangeManager,
+    bytes32 exchangeId,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountIn,
+    uint256 amountOutMin
+  ) external returns (uint256 amountOut) {
+    require(isExchangeManager[exchangeManager], "ExchangeManager does not exist");
+    amountOut = IExchangeManager(exchangeManager).swapIn(exchangeId, tokenIn, tokenOut, amountIn);
```

Comment:
So my brain was a bit absent on Friday, and I realized that I forgot the most important thing 😂 
The Broker's main responsibility is to actually execute the token transfers/mint/burn involved in a swap.
The `IExchangeManager`'s only responsible for pricing things. Check [this section](https://docs.google.com/document/d/1fII7pHpBqnlPtDXj0JdvAIAr6V5JZN5ZuJREYYCaww8/edit#heading=h.wquz253tdlws) of the design doc.

So the broker should have two internal functions:
1. `transferOut(address to, address token,  uint256 amount)` 
2. `transferIn(address token, uint256 amount)`

Each will have to split logic based on what type of token they're dealing with, like so:

**transferOut**

- if `Reserve.isStableAsset(token)` then `StableToken(token).mint(to, amount)`.
- if `Reserve.isCollateralAsset(token)` then `Reserve.transferExchangeCollateralAsset(token, to, amount)`

**transferIn**

- if `Reserve.isStableAsset(token) then `StableToen(token).transferFrom(to, address(this), amount)` && `StableToken(token).burn(amount)`
- if `Reserve.isCollateralAsset(token)` then `IERC20(token).transferFrom(to, address(reserve), amount)`

---

Then basically you can use the functions in `swapIn` and `swapOut`, to execute the trade `transferIn(tokenIn, amountIn)` and `transferOut(tokenOut, amountOut)`.

Makes sense?